scripts/check_copyright_header/src/main.rs

```diff
@@ -0,0 +1,150 @@
+use std::{
+    env,
+    fs::File,
+    io::{BufRead, BufReader},
+};
+
+fn check_file_header(lines: impl IntoIterator<Item = String>) -> Result<(), &'static str> {
+    let mut found_zefchain_labs_header = false;
+    let mut is_end_of_header = false;
+
+    for line in lines {
+        if !is_end_of_header {
+            if line == "// Copyright (c) Zefchain Labs, Inc." {
+                found_zefchain_labs_header = true;
+                continue;
+            }
+
+            if line.starts_with("// Copyright (c)") {
+                continue;
+            }
+
+            if line == "// SPDX-License-Identifier: Apache-2.0" {
+                is_end_of_header = true;
+                continue;
+            }
+
+            return Err("Unexpected line reached");
+        } else {
+            if !found_zefchain_labs_header {
+                return Err("Incorrect copyright header, Zefchain Labs header not found");
+            }
+
+            if line.is_empty() {
+                // Found separation line
+                return Ok(());
+            } else {
+                return Err("Separation line not found");
+            }
+        }
+    }
+
+    Err("Incorrect copyright header")
+}
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+    let file_path = args.get(1).expect("Usage: FILE");
+
+    let file = File::open(file_path).expect("Failed to open file");
+    let reader = BufReader::new(file);
+
+    let lines: Vec<String> = reader
+        .lines()
+        .collect::<Result<_, _>>()
```

Comment:
Maybe change the signature of `check_file_header` to match `reader.lines()` instead of collecting?

Reply:
Actually if I do that I'll have to have an input type for the function like `Lines<BufReader<File>>`, and creating that for the tests will be a pain 😅 or at least I haven't found any easy way to do it.
Is there any major disadvantage of keeping it as is?

Reply:
I'm not sure if I understood correctly, but I think the goal is to have a signature with `lines: impl IntoIterator<Item = Result<String, io::Error>>`.